File: skills/instagram-publisher/scripts/publish.js
#!/usr/bin/env node
// Instagram Carousel Publisher
// Usage:
//   node --env-file=.env publish.js --images "slide1.jpg,slide2.jpg" --caption "..." [--dry-run]
//   node --env-file=.env publish.js --dir ./slides/jpeg --caption-file caption.txt [--dry-run]
//   node --env-file=.env publish.js --dir ./slides/jpeg --caption "..." [--dry-run]

import { readFileSync, readdirSync, statSync } from 'node:fs';
import { resolve, join, extname } from 'node:path';
import { fileURLToPath } from 'node:url';

// ── Argument parsing ──────────────────────────────────────────

export function parseArgs(argv) {
  const args = { images: [], caption: '', dryRun: false, dir: '' };
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === '--images') {
      if (i + 1 < argv.length) args.images = argv[++i].split(',').map(s => s.trim());
    }
    else if (argv[i] === '--dir') {
      if (i + 1 < argv.length) args.dir = argv[++i];
    }
    else if (argv[i] === '--caption') {
      if (i + 1 < argv.length) args.caption = argv[++i];
    }
    else if (argv[i] === '--caption-file') {
      if (i + 1 < argv.length) args.caption = readFileSync(argv[++i], 'utf-8').trim();
    }
    else if (argv[i] === '--dry-run') args.dryRun = true;
  }

  // If --dir provided and no --images, scan directory for images
  if (args.dir && !args.images.length) {
    const dir = resolve(args.dir);
    const exts = ['.jpg', '.jpeg', '.png'];
    args.images = readdirSync(dir)
      .filter(f => exts.includes(extname(f).toLowerCase()))
      .sort()
      .map(f => join(dir, f));
  }

  return args;
}

// ── Image upload (imgbb) ──────────────────────────────────────

export async function uploadToImgbb(imagePath, apiKey) {
  const absolutePath = resolve(imagePath);
  const base64 = readFileSync(absolutePath).toString('base64');
  const body = new URLSearchParams({ image: base64 });
  const res = await fetch(`https://api.imgbb.com/1/upload?key=${apiKey}`, {
    method: 'POST',
    body,
  });
  if (!res.ok) throw new Error(`imgbb upload failed [${res.status}]: ${await res.text()}`);
  const json = await res.json();
  return json.data.url;
}

// ── Instagram Graph API ───────────────────────────────────────

const IG_BASE = 'https://graph.facebook.com/v21.0';

export async function createChildContainer(userId, imageUrl, accessToken) {
  const params = new URLSearchParams({
    image_url: imageUrl,
    is_carousel_item: 'true',
    access_token: accessToken,
  });
  const res = await fetch(`${IG_BASE}/${userId}/media?${params}`, { method: 'POST' });
  if (!res.ok) throw new Error(`createChildContainer failed [${res.status}]: ${await res.text()}`);
  return (await res.json()).id;
}

export async function createSingleContainer(userId, imageUrl, caption, accessToken) {
  const params = new URLSearchParams({
    image_url: imageUrl,
    caption,
    access_token: accessToken,
  });
  const res = await fetch(`${IG_BASE}/${userId}/media?${params}`, { method: 'POST' });
  if (!res.ok) throw new Error(`createSingleContainer failed [${res.status}]: ${await res.text()}`);
  return (await res.json()).id;
}

export async function getContainerStatus(containerId, accessToken) {
  const params = new URLSearchParams({ fields: 'status_code', access_token: accessToken });
  const res = await fetch(`${IG_BASE}/${containerId}?${params}`);
  if (!res.ok) throw new Error(`getContainerStatus failed [${res.status}]: ${await res.text()}`);
  return (await res.json()).status_code;
}

export async function pollUntilFinished(containerId, accessToken, timeoutMs = 60_000, intervalMs = 3_000) {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    const status = await getContainerStatus(containerId, accessToken);
    if (status === 'FINISHED') return;
    if (status === 'ERROR') throw new Error(`Container ${containerId} entered ERROR state`);
    await new Promise(r => setTimeout(r, intervalMs));
  }
  throw new Error(`Container ${containerId} timed out after ${timeoutMs}ms`);
}

export async function createCarouselContainer(userId, childIds, caption, accessToken) {
  const params = new URLSearchParams({
    media_type: 'CAROUSEL_ALBUM',
    children: childIds.join(','),
    caption,
    access_token: accessToken,
  });
  const res = await fetch(`${IG_BASE}/${userId}/media?${params}`, { method: 'POST' });
  if (!res.ok) throw new Error(`createCarouselContainer failed [${res.status}]: ${await res.text()}`);
  return (await res.json()).id;
}

export async function publishMedia(userId, containerId, accessToken) {
  const params = new URLSearchParams({ creation_id: containerId, access_token: accessToken });
  const res = await fetch(`${IG_BASE}/${userId}/media_publish?${params}`, { method: 'POST' });
  if (!res.ok) throw new Error(`publishMedia failed [${res.status}]: ${await res.text()}`);
  return (await res.json()).id;
}

export async function getPermalink(mediaId, accessToken) {
  const params = new URLSearchParams({ fields: 'permalink', access_token: accessToken });
  const res = await fetch(`${IG_BASE}/${mediaId}?${params}`);
  if (!res.ok) return null;
  const json = await res.json();
  return json.permalink ?? null;
}

// ── Main ──────────────────────────────────────────────────────

async function main() {
  const { images, caption, dryRun } = parseArgs(process.argv);

  if (!images.length) throw new Error('--images or --dir is required');
  if (!caption) throw new Error('--caption or --caption-file is required');
  if (images.length > 10) {
    throw new Error(`Instagram allows max 10 images per post (got ${images.length})`);
  }
  if (caption.length > 2200) {
    throw new Error(`Caption exceeds Instagram's 2200-character limit (got ${caption.length})`);
  }

  const { IMGBB_API_KEY, INSTAGRAM_ACCESS_TOKEN, INSTAGRAM_USER_ID } = process.env;
  if (!IMGBB_API_KEY) throw new Error('IMGBB_API_KEY is not set in environment');
  if (!INSTAGRAM_ACCESS_TOKEN) throw new Error('INSTAGRAM_ACCESS_TOKEN is not set in environment');
  if (!INSTAGRAM_USER_ID) throw new Error('INSTAGRAM_USER_ID is not set in environment');

  // Validate images exist
  for (const img of images) {
    try { statSync(resolve(img)); } catch {
      throw new Error(`Image not found: ${img}`);
    }
  }

  const isSingle = images.length === 1;
  const postType = isSingle ? 'single image' : `carousel (${images.length} slides)`;

  console.log(`📸 Uploading ${images.length} image(s) to imgbb...`);
  const imageUrls = [];
  for (let i = 0; i < images.length; i++) {
    const url = await uploadToImgbb(images[i], IMGBB_API_KEY);
    imageUrls.push(url);
    console.log(`   [${i + 1}] ${url}`);
  }

  let containerId;

  if (isSingle) {
    // Single image post
    console.log('\n📦 Creating Instagram media container...');
    containerId = await createSingleContainer(INSTAGRAM_USER_ID, imageUrls[0], caption, INSTAGRAM_ACCESS_TOKEN);
    console.log(`   Container ID: ${containerId}`);

    console.log('\n⏳ Waiting for container to process...');
    await pollUntilFinished(containerId, INSTAGRAM_ACCESS_TOKEN);
    console.log('   Container ready.');
  } else {
    // Carousel post (2-10 images)
    console.log('\n📦 Creating Instagram media containers...');
    const childIds = [];
    for (const url of imageUrls) {
      const id = await createChildContainer(INSTAGRAM_USER_ID, url, INSTAGRAM_ACCESS_TOKEN);
      childIds.push(id);
    }
    console.log(`   Container IDs: ${childIds.join(', ')}`);

    console.log('\n⏳ Waiting for containers to finish processing...');
    await Promise.all(childIds.map(id => pollUntilFinished(id, INSTAGRAM_ACCESS_TOKEN)));
    console.log('   All containers ready.');

    console.log('\n🎠 Creating carousel container...');
    containerId = await createCarouselContainer(
      INSTAGRAM_USER_ID, childIds, caption, INSTAGRAM_ACCESS_TOKEN
    );
    await pollUntilFinished(containerId, INSTAGRAM_ACCESS_TOKEN);
    console.log(`   Carousel container ID: ${containerId}`);
  }

  if (dryRun) {
    console.log(`\n✅ DRY RUN complete — ${postType} ready but not published.`);
    console.log(`   Container: ${containerId}`);
    return;
  }

  console.log('\n🚀 Publishing to Instagram...');
  const postId = await publishMedia(INSTAGRAM_USER_ID, containerId, INSTAGRAM_ACCESS_TOKEN);
  const permalink = await getPermalink(postId, INSTAGRAM_ACCESS_TOKEN);
  console.log(`\n✅ Published successfully!`);
  console.log(`   Type: ${postType}`);
  console.log(`   Post ID: ${postId}`);
  if (permalink) console.log(`   URL: ${permalink}`);
}

// Run only when executed directly (not when imported for tests)
const isMain = process.argv[1] === fileURLToPath(import.meta.url);
if (isMain) {
  main().catch(err => {
    console.error(`\n❌ ${err.message}`);
    process.exit(1);
  });
}

File: squads/instagram-carousel/output/2026-03-18-205435/v1/publish.py
#!/usr/bin/env python3
"""Instagram Carousel Publisher via curl subprocess calls."""
import subprocess, json, time, sys, os

IMGBB_KEY = "a85aca5816f0b6885e9fea5cba269f6b"
IG_TOKEN = os.environ.get("INSTAGRAM_ACCESS_TOKEN", "")
IG_USER = os.environ.get("INSTAGRAM_USER_ID", "")
IG_BASE = "https://graph.facebook.com/v21.0"
JPEG_DIR = "squads/instagram-carousel/output/2026-03-18-205435/v1/slides/rendered/jpeg"
DRY_RUN = "--dry-run" in sys.argv

CAPTION = """O que é DMRI seca? 90% das pessoas não sabem — e ela é a forma mais comum de degeneração macular.

A DMRI seca representa 9 em cada 10 casos de degeneração macular relacionada à idade. No Brasil, 3 milhões de pessoas convivem com essa condição.

Ela não dói. Não dá sintomas claros no início. E não tem cura.

Mas tem como retardar a progressão. E quanto mais cedo você descobre, mais visão preserva.

Nesse carrossel, reuni 5 fatos atualizados sobre a DMRI seca que podem mudar a forma como você cuida dos seus olhos:

- 90% dos casos de DMRI são da forma seca
- Suplementação AREDS retarda a progressão (não é qualquer vitamina)
- A visão central é afetada, mas a periférica é preservada
- Fotobiomodulação (luz, sem agulha) foi aprovada em 2024
- O teste de Amsler detecta mudanças em 30 segundos, em casa

Conhecimento é a melhor prevenção. Salve para consultar depois e mande para quem precisa saber.

Se faz tempo que você não fez um exame completo, agende sua avaliação. Link na bio.

Dr. Thiago Russo — CRM/AM
Oftalmologista — CEDOA Manaus

#dmri #saudeocular #oftalmologista #cedoamanaus #degeneracaomacular"""

def curl_post(url, form_data=None, data=None):
    cmd = ["curl", "-s", "--max-time", "60", "-X", "POST", url]
    if form_data:
        for k, v in form_data:
            cmd += ["--form", f"{k}={v}"]
    if data:
        for k, v in data:
            cmd += ["--data-urlencode", f"{k}={v}"]
    result = subprocess.run(cmd, capture_output=True, text=True)
    return json.loads(result.stdout)

def curl_get(url, params):
    qs = "&".join(f"{k}={v}" for k, v in params)
    full = f"{url}?{qs}"
    result = subprocess.run(["curl", "-s", "--max-time", "30", full], capture_output=True, text=True)
    return json.loads(result.stdout)

# Step 1: Upload images
print("📸 Uploading 9 images to imgbb...")
image_urls = []
for i in range(1, 10):
    filepath = f"{JPEG_DIR}/slide-{i:02d}.jpg"
    resp = curl_post("https://api.imgbb.com/1/upload", form_data=[
        ("key", IMGBB_KEY),
        ("image", f"@{filepath}"),
        ("expiration", "600"),
    ])
    url = resp["data"]["url"]
    image_urls.append(url)
    print(f"   [{i}] {url}")

# Step 2: Create child containers
print("\n📦 Creating Instagram media containers...")
child_ids = []
for url in image_urls:
    resp = curl_post(f"{IG_BASE}/{IG_USER}/media", data=[
        ("image_url", url),
        ("is_carousel_item", "true"),
        ("access_token", IG_TOKEN),
    ])
    if "id" not in resp:
        print(f"   ERROR: {resp}")
        sys.exit(1)
    child_ids.append(resp["id"])
print(f"   Container IDs: {', '.join(child_ids)}")

# Step 3: Poll containers
print("\n⏳ Waiting for containers to process...")
for cid in child_ids:
    for _ in range(20):
        resp = curl_get(f"{IG_BASE}/{cid}", [("fields", "status_code"), ("access_token", IG_TOKEN)])
        status = resp.get("status_code", "UNKNOWN")
        if status == "FINISHED":
            break
        if status == "ERROR":
            print(f"   ERROR on container {cid}")
            sys.exit(1)
        time.sleep(3)
print("   All containers ready.")

# Step 4: Create carousel
print("\n🎠 Creating carousel container...")
children_csv = ",".join(child_ids)
resp = curl_post(f"{IG_BASE}/{IG_USER}/media", data=[
    ("media_type", "CAROUSEL_ALBUM"),
    ("children", children_csv),
    ("caption", CAPTION),
    ("access_token", IG_TOKEN),
])
if "id" not in resp:
    print(f"   ERROR: {resp}")
    sys.exit(1)
carousel_id = resp["id"]
print(f"   Carousel ID: {carousel_id}")

# Poll carousel
for _ in range(20):
    resp = curl_get(f"{IG_BASE}/{carousel_id}", [("fields", "status_code"), ("access_token", IG_TOKEN)])
    if resp.get("status_code") == "FINISHED":
        break
    time.sleep(3)

if DRY_RUN:
    print("\n✅ DRY RUN complete — carousel ready but not published.")
    print(f"   Carousel container: {carousel_id}")
    sys.exit(0)

# Step 5: Publish
print("\n🚀 Publishing to Instagram...")
resp = curl_post(f"{IG_BASE}/{IG_USER}/media_publish", data=[
    ("creation_id", carousel_id),
    ("access_token", IG_TOKEN),
])
if "id" not in resp:
    print(f"   ERROR: {resp}")
    sys.exit(1)
post_id = resp["id"]

# Get permalink
resp = curl_get(f"{IG_BASE}/{post_id}", [("fields", "permalink"), ("access_token", IG_TOKEN)])
permalink = resp.get("permalink", "N/A")

print(f"\n✅ Published successfully!")
print(f"   Post ID: {post_id}")
print(f"   URL: {permalink}")
